File: internal/protocol/apps.go
package protocol

import "encoding/json"

const (
	CommandAppsList            = "apps.list"
	CommandAppsPackagePreview  = "apps.package_preview"
	CommandAppsPackageActivate = "apps.package_activate"
	CommandAppsConfigStatus    = "apps.config_status"
	CommandAppsConfigApply     = "apps.config_apply"
	CommandAppsInvoke          = "apps.invoke"

	AppSchemaVersion = "hank.app.v1"
	AppRuntimeStdio  = "stdio"

	AppUserAccessAdminsOnly  = "admins_only"
	AppUserAccessHomeMembers = "home_members"
)

type AppSummary struct {
	ID              string              `json:"id"`
	Name            string              `json:"name"`
	Version         string              `json:"version"`
	Publisher       string              `json:"publisher,omitempty"`
	Description     string              `json:"description,omitempty"`
	Enabled         bool                `json:"enabled"`
	Status          string              `json:"status"`
	LastError       string              `json:"last_error,omitempty"`
	Capabilities    []string            `json:"capabilities,omitempty"`
	SlashCommands   []AppSlashCommand   `json:"slash_commands,omitempty"`
	Commands        []AppCommandSummary `json:"commands,omitempty"`
	SettingsSchema  AppSettingsSchema   `json:"settings_schema,omitempty"`
	UserAccess      string              `json:"user_access,omitempty"`
	PublicConfig    json.RawMessage     `json:"public_config,omitempty"`
	SecretFieldsSet map[string]bool     `json:"secret_fields_set,omitempty"`
}

type AppSlashCommand struct {
	Command     string `json:"command"`
	CommandID   string `json:"command_id"`
	Description string `json:"description,omitempty"`
}

type AppCommandSummary struct {
	ID             string `json:"id"`
	Mode           string `json:"mode"`
	TimeoutSeconds int    `json:"timeout_seconds"`
	AdminOnly      bool   `json:"admin_only"`
}

type AppSettingsSchema struct {
	Fields []AppSettingsField `json:"fields,omitempty"`
}

type AppSettingsField struct {
	Key         string              `json:"key"`
	Label       string              `json:"label,omitempty"`
	Type        string              `json:"type"`
	Required    bool                `json:"required,omitempty"`
	Secret      bool                `json:"secret,omitempty"`
	SecretKey   string              `json:"secret_key,omitempty"`
	Source      string              `json:"source,omitempty"`
	Placeholder string              `json:"placeholder,omitempty"`
	Help        string              `json:"help,omitempty"`
	Default     json.RawMessage     `json:"default,omitempty"`
	Order       int                 `json:"order,omitempty"`
	Options     []AppSettingsOption `json:"options,omitempty"`
}

type AppSettingsOption struct {
	Value json.RawMessage `json:"value"`
	Label string          `json:"label,omitempty"`
}

type AppsListRequest struct{}

type AppsListResponse struct {
	Apps []AppSummary `json:"apps"`
}

type AppsPackagePreviewRequest struct {
	StagingID     string `json:"staging_id"`
	DownloadURL   string `json:"download_url"`
	DownloadToken string `json:"download_token"`
}

type AppsPackagePreviewResponse struct {
	StagingID string     `json:"staging_id"`
	App       AppSummary `json:"app"`
	Warnings  []string   `json:"warnings,omitempty"`
	Replacing bool       `json:"replacing"`
}

type AppsPackageActivateRequest struct {
	StagingID string `json:"staging_id"`
	Enable    bool   `json:"enable"`
}

type AppsPackageActivateResponse struct {
	App AppSummary `json:"app"`
}

type AppsConfigStatusRequest struct {
	AppID string `json:"app_id,omitempty"`
}

type AppsConfigStatusResponse struct {
	Apps []AppSummary `json:"apps"`
}

type AppsConfigApplyRequest struct {
	AppID        string          `json:"app_id"`
	PublicConfig json.RawMessage `json:"public_config,omitempty"`
	Secrets      json.RawMessage `json:"secrets,omitempty"`
	Enable       *bool           `json:"enable,omitempty"`
	UserAccess   string          `json:"user_access,omitempty"`
}

type AppsConfigApplyResponse struct {
	App AppSummary `json:"app"`
}

type AppsInvokeRequest struct {
	AppID     string          `json:"app_id"`
	CommandID string          `json:"command_id"`
	Input     json.RawMessage `json:"input,omitempty"`
	Context   json.RawMessage `json:"context,omitempty"`
}

type AppsInvokeResponse struct {
	Output json.RawMessage `json:"output,omitempty"`
	JobID  string          `json:"job_id,omitempty"`
}

File: web/dashboard/src/api/apps.ts
import { apiClient, type ApiTransport } from "./client";
import { arrayFrom } from "./normalize";

export type AppSettingsField = {
  key: string;
  label?: string;
  type: "boolean" | "number" | "select" | "url" | "text" | string;
  required?: boolean;
  secret?: boolean;
  secret_key?: string;
  source?: string;
  placeholder?: string;
  help?: string;
  default?: unknown;
  order?: number;
  options?: Array<{ value: unknown; label?: string }>;
};

export type AppSummary = {
  id?: string;
  app_id?: string;
  name?: string;
  version?: string;
  publisher?: string;
  description?: string;
  enabled?: boolean;
  status?: string;
  last_error?: string;
  user_access?: "admins_only" | "home_members" | string;
  public_config?: Record<string, unknown> | string;
  public_config_json?: string;
  secret_fields_set?: Record<string, boolean> | string;
  secret_fields_set_json?: string;
  settings_schema?: { fields?: AppSettingsField[] } | string;
  settings_schema_json?: string;
  updated_at?: string;
};

export type AppsListPayload = {
  apps: AppSummary[];
};

export type AppsPackagePreview = {
  staging_id: string;
  package_sha256?: string;
  replacing?: boolean;
  warnings?: string[];
  app: AppSummary;
};

export type AppsActivateInput = {
  staging_id: string;
  package_sha256?: string;
  enable: boolean;
};

export type AppsConfigInput = {
  public_config?: Record<string, unknown>;
  secrets?: Record<string, unknown>;
  enable?: boolean;
  user_access?: string;
};

export class AppsClient {
  constructor(private readonly api: ApiTransport = apiClient) {}

  async listApps(): Promise<AppsListPayload> {
    const payload = await this.api.request<Partial<AppsListPayload>>("/v1/home/apps");
    return { apps: arrayFrom<AppSummary>(payload.apps) };
  }

  previewPackage(formData: FormData) {
    return this.api.request<AppsPackagePreview>("/v1/home/apps/import/preview", {
      method: "POST",
      body: formData,
    });
  }

  activatePackage(input: AppsActivateInput) {
    return this.api.request<{ app: AppSummary }>("/v1/home/apps/import/activate", {
      method: "POST",
      body: input,
    });
  }

  saveConfig(appID: string, input: AppsConfigInput) {
    return this.api.request<{ app: AppSummary }>(`/v1/home/apps/${encodeURIComponent(appID)}/config`, {
      method: "PUT",
      body: input,
    });
  }
}

export const appsClient = new AppsClient();
